### ISSUE TITLE
DNS Provider for CloudStack

### ISSUE THREAD

**Issue by wido** [apache/cloudstack#9958]
##### ISSUE TYPE
 * Feature Idea

##### COMPONENT NAME
~~~
API
Management Server
~~~

##### CLOUDSTACK VERSION
~~~
main
~~~

##### SUMMARY
My idea is as follows:

- A new framework is addedd: Authorative DNS
- This allows for admins and end-users to manage DNS zones
- Different providers implement different APIs, examples:
   - PowerDNS through it's API [1]
   - Public DNS providers like NS1 [2], rcodezero [3], Gandi [4], CloudFlare [5], etc
   - Your local DNS which applies to your environment

These drivers will support various functionality and advertise their capabilities or limitations:

- Create and delete zones
- Manage records in these zones
- Advertise which record types are supported
- Limit the amount of records in a zone (if any)
- Set a limitation on how low the TTL can be

Per account and domain we can then set limits:

- Amount of zones to create
- Amount of creates per zone

This would allow for a couple of things:

- End-users can manage their DNS via the same API as they manage their cloud resources
- End-users can manage DNS via the CloudStack UI
- We can directly create public DNS records for newly created instances
  - You assign a DNS zone to a CloudStack zone
  - Upon VM creation a record is created
    i-xx-yy-vm.myzone.tld A 1.2.3.4
    i-xx-yy-vm.myzone.tld AAAA 2001:db8:af31::61
  - When the VM is removed the DNS records are removed
- Console Proxy and Secondary Storage VMs can point to a working hostname instead of IP-address
  - We can then also support IPv6 for the CP and SS

**Comment by rohityadavcloud** [comment]
+1 a framework - plugin model would be great!

**Comment by kasimon** [comment]
Maybe some inspiration can be taken from OpenStack's Designate component, which serves a similar purpose.

**Comment by weizhouapache** [comment]
good idea to have a DNS framework !


PS: it looks ACS supports globodns (https://github.com/apache/cloudstack/tree/main/plugins/network-elements/globodns). however I doubt if it works ..

**Comment by wido** [comment]
> good idea to have a DNS framework !
> 
> PS: it looks ACS supports globodns (https://github.com/apache/cloudstack/tree/main/plugins/network-elements/globodns). however I doubt if it works ..

Yes, but that has a very limited scope. I think that as @kasimon writes, something like OpenStack's Designate is something we should want. Let me look into that.

I will work on this for the coming months, but I don't expect any results prior to summer 2025 (Time restricted).

**Comment by rohityadavcloud** [comment]
@wido have you done or started any work on this yet?

**Comment by wido** [comment]
> [@wido](https://github.com/wido) have you done or started any work on this yet?

Yes, slowly I'm working on this one.

**Comment by DaanHoogland** [comment]
@wido, is it ok to change this to an idea-discussion-item? it will remain a valid feature but as of yet seems largely an idea for a  distant future improvement and not a direct issue that can be picked up for implementation.

**Comment by wido** [comment]
> [@wido](https://github.com/wido), is it ok to change this to an idea-discussion-item? it will remain a valid feature but as of yet seems largely an idea for a distant future improvement and not a direct issue that can be picked up for implementation.

Yes and No. I discussed with @weizhouapache in Milan and if I am not mistaken he wants to start working on this in Q1 2026.

**Comment by weizhouapache** [comment]
> > [@wido](https://github.com/wido), is it ok to change this to an idea-discussion-item? it will remain a valid feature but as of yet seems largely an idea for a distant future improvement and not a direct issue that can be picked up for implementation.
> 
> Yes and No. I discussed with [@weizhouapache](https://github.com/weizhouapache) in Milan and if I am not mistaken he wants to start working on this in Q1 2026.

Yes, I've written a draft of design, can you review ?
https://cwiki.apache.org/confluence/display/CLOUDSTACK/DNS+Framework+and+Plugins

@wido @alexandremattioli @DaanHoogland @rohityadavcloud @sudo87

**Comment by wido** [comment]
> > > [@wido](https://github.com/wido), is it ok to change this to an idea-discussion-item? it will remain a valid feature but as of yet seems largely an idea for a distant future improvement and not a direct issue that can be picked up for implementation.
> > 
> > 
> > Yes and No. I discussed with [@weizhouapache](https://github.com/weizhouapache) in Milan and if I am not mistaken he wants to start working on this in Q1 2026.
> 
> Yes, I've written a draft of design, can you review ? https://cwiki.apache.org/confluence/display/CLOUDSTACK/DNS+Framework+and+Plugins
> 
> [@wido](https://github.com/wido) [@alexandremattioli](https://github.com/alexandremattioli) [@DaanHoogland](https://github.com/DaanHoogland) [@rohityadavcloud](https://github.com/rohityadavcloud) [@sudo87](https://github.com/sudo87)

Thank you! This has landed somewhere below in my mailbox, but I have taken a look at your proposal and I think overall it's solid. I do have a few remarks:

- Should we store the record contents locally? How do we sync this? Or should we just use the DNS API to fetch all the records and show the zone contents? What if somebody makes a change on the DNS server's end without telling CS?  The zone will get out of sync
- Maybe we should just store the zone name and an external reference, this could be an ID, UUID, Zone name, etc
- Shouldn't zones be unique per DNS server?

The DNS server is the Authoritative one here, so should we re-invent the wheel? We could always cache the response locally, but for consistency sake depend on the DNS server.

**Comment by weizhouapache** [comment]
thanks @wido for the review

> * Should we store the record contents locally? How do we sync this? Or should we just use the DNS API to fetch all the records and show the zone contents? What if somebody makes a change on the DNS server's end without telling CS?  The zone will get out of sync

if we do not store the DNS records in database, it will be more simple.
the side-effect is, for each DNS record operation (CRUD), we need to get remote DNS records at first. since the operations are not latency-sensitive, I think it fine to not store them.
therefore I  have removed the definition of `dns_record` table from the design doc.

> * Maybe we should just store the zone name and an external reference, this could be an ID, UUID, Zone name, etc
> * Shouldn't zones be unique per DNS server?
> 
> The DNS server is the Authoritative one here, so should we re-invent the wheel? We could always cache the response locally, but for consistency sake depend on the DNS server.

For clarification, I renamed `dns_server` table to `dns_server_account` which represents the accounts on external DNS servers. For each account on external DNS server, they can manage multiple DNS zones/domains.

the `dns_zone` table stores the information of DNS zones/domains, as well as some cloudstack information, for example ownership or dedication. I added a column `external_reference` to store the additional information of the dns zone on external dns servers.

I added a new table `dns_zone_network_map`. Users can Link the zone to multiple networks for auto-registration. user can specify the sub domain (optional)

cc @sudo87 @rajujith

**Comment by wido** [comment]
> thanks [@wido](https://github.com/wido) for the review
> 
> > * Should we store the record contents locally? How do we sync this? Or should we just use the DNS API to fetch all the records and show the zone contents? What if somebody makes a change on the DNS server's end without telling CS?  The zone will get out of sync
> 
> if we do not store the DNS records in database, it will be more simple. the side-effect is, for each DNS record operation (CRUD), we need to get remote DNS records at first. since the operations are not latency-sensitive, I think it fine to not store them. therefore I have removed the definition of `dns_record` table from the design doc.
> 

It will add latency. If this really becomes a problem we can potentially cache, but I wouldn't do that for now. Usually these APIs are rather fast.

> > * Maybe we should just store the zone name and an external reference, this could be an ID, UUID, Zone name, etc
> > * Shouldn't zones be unique per DNS server?
> > 
> > The DNS server is the Authoritative one here, so should we re-invent the wheel? We could always cache the response locally, but for consistency sake depend on the DNS server.
> 
> For clarification, I renamed `dns_server` table to `dns_server_account` which represents the accounts on external DNS servers. For each account on external DNS server, they can manage multiple DNS zones/domains.
> 

Does the "account" suffix make it any clearer? Because it's a generic DNS server. Usually these have a single API key/secret and you manage all the zones. CloudStack will then do the logic of accounting and authorization for it's internal users.

- dns_server
- dns_provider

Both would work for me.

> the `dns_zone` table stores the information of DNS zones/domains, as well as some cloudstack information, for example ownership or dedication. I added a column `external_reference` to store the additional information of the dns zone on external dns servers.
> 

The "external_reference" column would have to be populated by the driver. The driver would return this reference once the zone is created. That's what we then use to look up the zone again in that driver. The driver might return the zone name, an ID (integer) or a UUID. That's up to the driver.

> I added a new table `dns_zone_network_map`. Users can Link the zone to multiple networks for auto-registration. user can specify the sub domain (optional)
> 
> cc [@sudo87](https://github.com/sudo87) [@rajujith](https://github.com/rajujith)

I would also make sure that from the start there is DNSSEC support. That you can enable/disable DNSSEC by asking this and the provider can return the KSK and ZSK from the zone.

**Comment by weizhouapache** [comment]
> It will add latency. If this really becomes a problem we can potentially cache, but I wouldn't do that for now. Usually these APIs are rather fast.

let's not cache them then.

> 
> Does the "account" suffix make it any clearer? Because it's a generic DNS server. Usually these have a single API key/secret and you manage all the zones. CloudStack will then do the logic of accounting and authorization for it's internal users.
> 
> * dns_server
> * dns_provider
> 
> Both would work for me.

for internal cloud platform, if all vms use the same DNS provider with same credential, it is simple.

if we consider a multi-tenant platform where each domain or project represents a customer, it might be useful if each customer manages multiple accounts on various DNS providers/servers. Each account manages multiple DNS zone/domains.

> 
> > the `dns_zone` table stores the information of DNS zones/domains, as well as some cloudstack information, for example ownership or dedication. I added a column `external_reference` to store the additional information of the dns zone on external dns servers.
> 
> The "external_reference" column would have to be populated by the driver. The driver would return this reference once the zone is created. That's what we then use to look up the zone again in that driver. The driver might return the zone name, an ID (integer) or a UUID. That's up to the driver.

Yes, make sense. that's what I intend to do: save necessary information in database to speed up operations afterwards.

> 
> I would also make sure that from the start there is DNSSEC support. That you can enable/disable DNSSEC by asking this and the provider can return the KSK and ZSK from the zone.

we will consider the use case.

**Comment by wido** [comment]
> > It will add latency. If this really becomes a problem we can potentially cache, but I wouldn't do that for now. Usually these APIs are rather fast.
> 
> let's not cache them then.
> 
> > Does the "account" suffix make it any clearer? Because it's a generic DNS server. Usually these have a single API key/secret and you manage all the zones. CloudStack will then do the logic of accounting and authorization for it's internal users.
> > 
> > * dns_server
> > * dns_provider
> > 
> > Both would work for me.
> 
> for internal cloud platform, if all vms use the same DNS provider with same credential, it is simple.
> 
> if we consider a multi-tenant platform where each domain or project represents a customer, it might be useful if each customer manages multiple accounts on various DNS providers/servers. Each account manages multiple DNS zone/domains.
> 

Yes, but we should also consider this case: Public cloud provider offering VMs. They have their own PowerDNS environment behind it. Customers can use CloudStack to:

- Deploy Virtual Machines
- Manage Object Storage
- Managed DNS Zones and Records

A single API to talk to for the customer. They have no idea what DNS environment is behind it, they just create a zone and that's it.

Come to think, I think that per provider/driver you should also be able to set the NS records. If set, these will be returned by CloudStack as the NS records people can use

dns_server_nameservers
- dns_server_id
- hostname

> > > the `dns_zone` table stores the information of DNS zones/domains, as well as some cloudstack information, for example ownership or dedication. I added a column `external_reference` to store the additional information of the dns zone on external dns servers.
> > 
> > 
> > The "external_reference" column would have to be populated by the driver. The driver would return this reference once the zone is created. That's what we then use to look up the zone again in that driver. The driver might return the zone name, an ID (integer) or a UUID. That's up to the driver.
> 
> Yes, make sense. that's what I intend to do: save necessary information in database to speed up operations afterwards.
> 
> > I would also make sure that from the start there is DNSSEC support. That you can enable/disable DNSSEC by asking this and the provider can return the KSK and ZSK from the zone.
> 
> we will consider the use case.

**Comment by weizhouapache** [comment]
> > > Does the "account" suffix make it any clearer? Because it's a generic DNS server. Usually these have a single API 
> Yes, but we should also consider this case: Public cloud provider offering VMs. They have their own PowerDNS environment behind it. Customers can use CloudStack to:
> 
> * Deploy Virtual Machines
> * Manage Object Storage
> * Managed DNS Zones and Records
> 
> A single API to talk to for the customer. They have no idea what DNS environment is behind it, they just create a zone and that's it.
> 

good use case.

I added two columns to `dns_server` table

- is_public (BOOLEAN) - Whether the DNS server is publicly accessible by other accounts without credentials.
- public_domain_suffix (VARCHAR(255)) - The domain suffix can be used by other accounts without credentials.

it can work in two scenaraios
- admin creates DNS server without credentials and `is_public` is `false`, users create DNS zones with their own credentials
- admin creates DNS server with credentials and `is_public` is `true`, users create DNS zones without specifying credentials, but the domain suffix is restricted.

how about it ?

> Come to think, I think that per provider/driver you should also be able to set the NS records. If set, these will be returned by CloudStack as the NS records people can use
> 
> dns_server_nameservers
> 
> * dns_server_id
> * hostname
> 

it may be useful for Public DNS server (`is_public` is `true` in the above cases)
CloudStack just needs to store the NS and returns to users, right ?

> > > > the `dns_zone` table stores the information of DNS zones/domains, as well as some cloudstack information, for example ownership or dedication. I added a column `external_reference` to store the additional information of the dns zone on external dns servers.
> > > 
> > > 
> > > The "external_reference" column would have to be populated by the driver. The driver would return this reference once the zone is created. That's what we then use to look up the zone again in that driver. The driver might return the zone name, an ID (integer) or a UUID. That's up to the driver.
> > 
> > 
> > Yes, make sense. that's what I intend to do: save necessary information in database to speed up operations afterwards.
> > > I would also make sure that from the start there is DNSSEC support. That you can enable/disable DNSSEC by asking this and the provider can return the KSK and ZSK from the zone.
> > 
> > 
> > we will consider the use case.

**Comment by wido** [comment]
> > > > Does the "account" suffix make it any clearer? Because it's a generic DNS server. Usually these have a single API
> > > > Yes, but we should also consider this case: Public cloud provider offering VMs. They have their own PowerDNS environment behind it. Customers can use CloudStack to:
> > 
> > 
> > 
> > * Deploy Virtual Machines
> > * Manage Object Storage
> > * Managed DNS Zones and Records
> > 
> > A single API to talk to for the customer. They have no idea what DNS environment is behind it, they just create a zone and that's it.
> 
> good use case.
> 
> I added two columns to `dns_server` table
> 
>     * is_public (BOOLEAN) - Whether the DNS server is publicly accessible by other accounts without credentials.
> 
>     * public_domain_suffix (VARCHAR(255)) - The domain suffix can be used by other accounts without credentials.
> 
> 
> it can work in two scenaraios
> 
>     * admin creates DNS server without credentials and `is_public` is `false`, users create DNS zones with their own credentials
> 
>     * admin creates DNS server with credentials and `is_public` is `true`, users create DNS zones without specifying credentials, but the domain suffix is restricted.
> 
> 
> how about it ?

Yes! But to check, what does the public_domain_suffix do exactly? I'm trying to understand that use-case.

> 
> > Come to think, I think that per provider/driver you should also be able to set the NS records. If set, these will be returned by CloudStack as the NS records people can use
> > dns_server_nameservers
> > 
> > * dns_server_id
> > * hostname
> 
> it may be useful for Public DNS server (`is_public` is `true` in the above cases) CloudStack just needs to store the NS and returns to users, right ?

Yes. But not always store them, just return them. Often users can't change the NS-records. So they create a new zone, the NS records are set (not always transparant) and we just need to return to the user what to use as NS-records.

> 
> > > > > the `dns_zone` table stores the information of DNS zones/domains, as well as some cloudstack information, for example ownership or dedication. I added a column `external_reference` to store the additional information of the dns zone on external dns servers.
> > > > 
> > > > 
> > > > The "external_reference" column would have to be populated by the driver. The driver would return this reference once the zone is created. That's what we then use to look up the zone again in that driver. The driver might return the zone name, an ID (integer) or a UUID. That's up to the driver.
> > > 
> > > 
> > > Yes, make sense. that's what I intend to do: save necessary information in database to speed up operations afterwards.
> > > > I would also make sure that from the start there is DNSSEC support. That you can enable/disable DNSSEC by asking this and the provider can return the KSK and ZSK from the zone.
> > > 
> > > 
> > > we will consider the use case.

**Comment by weizhouapache** [comment]
> Yes! But to check, what does the public_domain_suffix do exactly? I'm trying to understand that use-case.
> 

for security reason, if users do not need to pass the credentials, it would be safer to restrict the domain suffix they can use.
otherwise they might create unexpected DNS zones

> Yes. But not always store them, just return them. Often users can't change the NS-records. So they create a new zone, the NS records are set (not always transparant) and we just need to return to the user what to use as NS-records.

ok, I think NS records can be specified when add the DNS server , right ?
